File: src/main/java/engine/rendering/instances/renderers/entity/EntityRenderer.java
package engine.rendering.instances.renderers.entity;

import engine.architecture.scene.SceneContext;
import engine.architecture.scene.entity.Entity;
import engine.architecture.scene.node.Node;
import engine.rendering.abstracted.renderers.Renderer3D;
import engine.utils.libraryBindings.maths.joml.FrustumIntersection;
import engine.utils.libraryBindings.maths.joml.Vector3f;
import engine.utils.libraryBindings.maths.utils.Matrix4;
import engine.utils.libraryBindings.opengl.shaders.*;
import engine.utils.libraryBindings.opengl.textures.ITexture;
import lombok.Getter;
import lombok.var;

@Getter
public class EntityRenderer extends Renderer3D<Entity> {

    private final ShadersProgram<Entity> shadersProgram;
    private final FrustumIntersection frustumIntersection = new FrustumIntersection();

    private static final String VERT_FILE = "res/shaders/entity/entity_vs.glsl";
    private static final String FRAG_FILE = "res/shaders/entity/entity_fs.glsl";

    @Getter
    public static final EntityRenderer instance;

    static {
        try {
            instance = new EntityRenderer();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private EntityRenderer() throws Exception {
        this.shadersProgram = ShadersProgram.create(VERT_FILE, FRAG_FILE);

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<Entity>("albedoConst") {
            @Override
            public Vector3f getUniformValue(RenderState<Entity> state) {
                return new Vector3f(
                        state.getMesh().getMaterial().getDiffuseColor().x,
                        state.getMesh().getMaterial().getDiffuseColor().y,
                        state.getMesh().getMaterial().getDiffuseColor().z
                );
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<Entity>("albedoTex", 0) {
            @Override
            public ITexture getUniformValue(RenderState<Entity> state) {
                return state.getMesh().getMaterial().getDiffuseTexture();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<Entity>("isAlbedoMapped") {
            @Override
            public boolean getUniformValue(RenderState<Entity> state) {
                return state.getMesh().getMaterial().hasTexture();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<Entity>("normalMap", 1) {
            @Override
            public ITexture getUniformValue(RenderState<Entity> state) {
                return state.getMesh().getMaterial().getNormalTexture();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<Entity>("map_normal") {
            @Override
            public boolean getUniformValue(RenderState<Entity> state) {
                return state.getMesh().getMaterial().isUseNormalTex();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<Entity>("projectionMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<Entity> state) {
                return getContext().getCamera().getProjectionMatrix();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<Entity>("modelMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<Entity> state) {
                return state.getInstance().getTransform().getTransformationMatrix();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<Entity>("viewMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<Entity> state) {
                return getContext().getCamera().getViewMatrix();
            }
        });
    }

    @Override
    public void render(SceneContext context) {
        if (renderList.isEmpty()) {
            return;
        }
        shadersProgram.bind();

        RenderState<Entity> renderState = new RenderState<>(this, context.getCamera());
        shadersProgram.updatePerRenderUniforms(renderState);

        for (var model : renderList.keySet()) {
            for (int i = 0; i < model.getMeshes().length; i++) {
                model.bindAndConfigure(i);
                for (var entity : renderList.get(model)) {
                    frustumIntersection.set(
                            context.getCamera().getProjectionViewMatrix().mul(
                                    entity.getTransform().getTransformationMatrix(), Matrix4.pool.poolAndGive()
                            )
                    );
                    if (!checkRenderPass(entity)) {
                        continue;
                    }
                    RenderState<Entity> instanceState = new RenderState<>(this, entity, context.getCamera(), i);
                    shadersProgram.updatePerInstanceUniforms(instanceState);
                    model.render(instanceState, i);
                }
            }
        }

        shadersProgram.unbind();
    }

    @Override
    public void render(SceneContext context, Node.Condition condition) {
        if (renderList.isEmpty()) {
            return;
        }
        shadersProgram.bind();

        RenderState<Entity> renderState = new RenderState<>(this, context.getCamera());
        shadersProgram.updatePerRenderUniforms(renderState);

        for (var model : renderList.keySet()) {
            for (int i = 0; i < model.getMeshes().length; i++) {
                model.bindAndConfigure(i);
                for (var entity : renderList.get(model)) {
                    frustumIntersection.set(
                            context.getCamera().getProjectionViewMatrix().mul(
                                    entity.getTransform().getTransformationMatrix(), Matrix4.pool.poolAndGive()
                            )
                    );
                    if (!checkRenderPass(entity)) {
                        continue;
                    }
                    RenderState<Entity> instanceState = new RenderState<>(this, entity, context.getCamera(), i);
                    shadersProgram.updatePerInstanceUniforms(instanceState);
                    model.render(instanceState, i);
                }
            }
        }

        shadersProgram.unbind();
    }

    private boolean checkRenderPass(Entity entity) {
        return !checkClippingCulling(entity.getTransform().getPosition()) &&
                frustumIntersection.testAab(entity.getModel().getBounds().getMin(), entity.getModel().getBounds().getMax());
    }

    @Override
    public void cleanUp() {
        shadersProgram.delete();
    }
}

File: src/main/java/engine/rendering/instances/renderers/entity/EntityRenderer.kt
package engine.rendering.instances.renderers.entity

import engine.architecture.scene.SceneContext
import engine.architecture.scene.entity.Entity
import engine.architecture.scene.node.Node
import engine.rendering.abstracted.renderers.Renderer3D
import engine.utils.libraryBindings.maths.joml.FrustumIntersection
import engine.utils.libraryBindings.maths.joml.Vector3f
import engine.utils.libraryBindings.maths.utils.Matrix4
import engine.utils.libraryBindings.opengl.shaders.*
import engine.utils.libraryBindings.opengl.textures.ITexture

internal class EntityRenderer : Renderer3D<Entity>() {

    private val shadersProgram: ShadersProgram<Entity>
    private val frustumIntersection: FrustumIntersection = FrustumIntersection()

    init {
        this.shadersProgram = ShadersProgram.create(VERT_FILE, FRAG_FILE)

        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<Entity>("albedoConst") {
            override fun getUniformValue(state: RenderState<Entity>): Vector3f {
                return Vector3f(
                    state.mesh.material.diffuseColor.x,
                    state.mesh.material.diffuseColor.y,
                    state.mesh.material.diffuseColor.z
                )
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<Entity>("albedoTex", 0) {
            override fun getUniformValue(state: RenderState<Entity>): ITexture? {
                return state.mesh.material.diffuseTexture
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<Entity>("isAlbedoMapped") {
            override fun getUniformValue(state: RenderState<Entity>): Boolean {
                return state.mesh.material.hasTexture()
            }
        })

        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<Entity>("normalMap", 1) {
            override fun getUniformValue(state: RenderState<Entity>): ITexture? {
                return state.mesh.material.normalTexture
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<Entity>("map_normal") {
            override fun getUniformValue(state: RenderState<Entity>): Boolean {
                return state.mesh.material.isUseNormalTex
            }
        })


        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<Entity>("projectionMatrix") {
            override fun getUniformValue(state: RenderState<Entity>): UniformValue {
                return getContext().camera.projectionMatrix
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<Entity>("modelMatrix") {
            override fun getUniformValue(state: RenderState<Entity>): UniformValue {
                return state.instance.transform.transformationMatrix
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<Entity>("viewMatrix") {
            override fun getUniformValue(state: RenderState<Entity>): UniformValue {
                return getContext().camera.viewMatrix
            }
        })
    }

    override fun render(context: SceneContext) {
        if (renderList.size < 1) {
            return
        }
        shadersProgram.bind()

        val renderState = RenderState<Entity>(this, context.camera)
        shadersProgram.updatePerRenderUniforms(renderState)
        for (model in renderList.keys) {
            for (i in model.meshes.indices) {
                model.bindAndConfigure(i)
                for (entity in renderList[model]!!) {
                    frustumIntersection.set(
                        context.camera.projectionViewMatrix.mul(
                            entity.transform.transformationMatrix, Matrix4.pool.poolAndGive()
                        )
                    )
                    if (!checkRenderPass(entity) && entity.isActivated) {
                        continue
                    }
                    val instanceState = RenderState<Entity>(this, entity, context.camera, i)
                    shadersProgram.updatePerInstanceUniforms(instanceState)
                    model.render(instanceState, i)
                }
            }
        }

        shadersProgram.unbind()
    }

    override fun render(context: SceneContext, condition: Node.Condition) {
        if (renderList.size < 1) {
            return
        }
        shadersProgram.bind()

        val renderState = RenderState<Entity>(this, context.camera)
        shadersProgram.updatePerRenderUniforms(renderState)

        for (model in renderList.keys) {
            for (i in 0..model.meshes.size) {
                model.bindAndConfigure(i)
                for (entity in renderList[model]!!) {
                    frustumIntersection.set(
                        context.camera.projectionViewMatrix.mul(
                            entity.transform.transformationMatrix, Matrix4.pool.poolAndGive()
                        )
                    )
                    if (!checkRenderPass(entity) && entity.isActivated) {
                        continue
                    }
                    val instanceState = RenderState<Entity>(this, entity, context.camera, i)
                    shadersProgram.updatePerInstanceUniforms(instanceState)
                    model.render(instanceState, i)
                }
            }
        }

        shadersProgram.unbind()
    }

    private fun checkRenderPass(entity: Entity): Boolean {
        return !checkClippingCulling(entity.transform.position) &&
                frustumIntersection.testAab(entity.model.bounds.min, entity.model.bounds.max)
    }

    override fun cleanUp() {
        shadersProgram.delete()
    }

    companion object {

        private const val VERT_FILE = "res/shaders/entity/entity_vs.glsl"
        private const val FRAG_FILE = "res/shaders/entity/entity_fs.glsl"

        @JvmStatic
        val instance: EntityRenderer = EntityRenderer()

    }
}
